Move codespell configuration into pyproject.toml

yarikoptic opened the pull request:
So it is

- among other tools
- could be used/ran outside of pre-commit, e.g. just "codespell" or "codespell -w" to apply the fixes

henryiii commented:
Should we include tomli (maybe with a python version marker) in pre-commit? Also, do they allow the list to be a TOML list?

yarikoptic commented:
> Should we include tomli (maybe with a python version marker) in pre-commit? 

:facepalm:  . added now

> Also, do they allow the list to be a TOML list?

apparently (empirically tested) they do. Refactored

henryiii commented on .pre-commit-config.yaml:
```diff
@@ -33,5 +33,7 @@ repos:
-      args: ["-Ldynamc,notin"]
+      - id: codespell
+        additional_dependencies:
+          - tomli  # for python_version < '3.11'
```
```suggestion
          - "tomli; python_version<'3.11'"
```

95% sure this works, this is just passed on to the install tool.

henryiii commented:
Thanks! I'll have to update https://learn.scientific-python.org/development/guides/style/#spelling, which apparently does mention this, but I think it should be the recommendation now. :)

henryiii commented:
Though you still can't run this locally due to it picking up folders like `.nox`, `.mypy_cache`, and `.venv`. It would be nice if it could pick up on gitignores. I think we should either expand the list (eventually basically duplicating gitignore), or leave off hidden file checking. Most misspellings won't be in a hidden file.

```console
$ uvx codespell -w
Installed 1 package in 7ms
FIXED: ./tests/test_specifiers.py
FIXED: ./.nox/lint/lib/python3.9/site-packages/packaging/metadata.py
FIXED: ./.nox/lint/lib/python3.9/site-packages/readme_renderer/rst.py
FIXED: ./.nox/lint/lib/python3.9/site-packages/rfc3986-2.0.0.dist-info/METADATA
./.nox/lint/lib/python3.9/site-packages/pygments/unistring.py:40: Nd ==> And, 2nd
./.nox/lint/lib/python3.9/site-packages/pygments/lexers/_csound_builtins.py:54: endin ==> ending, end in
./.nox/lint/lib/python3.9/site-packages/pygments/lexers/_csound_builtins.py:1375: shiftin ==> shifting, shift in
./.nox/lint/lib/python3.9/site-packages/pygments/lexers/_csound_builtins.py:1412: soundin ==> sounding, sound in
...
./.mypy_cache/3.10/types.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/types.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/types.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/types.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/typing.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/typing.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/typing.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/typing.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/typing.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/typing.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/typing.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.10/_typeshed/__init__.data.json:1: trU ==> through, true
./.mypy_cache/3.12/inspect.data.json:1: ND ==> AND, 2ND
./.mypy_cache/3.12/inspect.data.json:1: ND ==> AND, 2ND
./.mypy_cache/3.12/inspect.data.json:1: ND ==> AND, 2ND
./.mypy_cache/3.12/types.data.json:1: asend ==> ascend, as end
./.mypy_cache/3.12/types.data.json:1: asend ==> ascend, as end
...
```

yarikoptic commented:
yeah, disabling `check-hidden ` would be the easiest way forward for now, or just always use via pre-commit which would IIRC limit only to git controlled files... there is also
- https://github.com/codespell-project/codespell/issues/1887

and workaround there is analogous to pre-commit invocation

```
codespell $(git ls-files)
```
